File: services/ark-api/ark-api/src/ark_api/models/tool_approvals.py
"""Models for ToolApprovalRequest API endpoints."""
from datetime import datetime
from typing import Optional, List, Any
from pydantic import BaseModel, Field


class ToolCallAnnotations(BaseModel):
    destructiveHint: Optional[bool] = None
    readOnlyHint: Optional[bool] = None
    idempotentHint: Optional[bool] = None
    openWorldHint: Optional[bool] = None


class ToolCallInfo(BaseModel):
    id: str
    name: str
    type: str
    arguments: str
    description: Optional[str] = None
    annotations: Optional[ToolCallAnnotations] = None
    agentReasoning: Optional[str] = None


class ExecutionContext(BaseModel):
    conversationHistory: str
    pendingToolCallIndex: int
    completedToolResults: Optional[List[str]] = None
    agentName: str
    agentNamespace: str


class QueryReference(BaseModel):
    name: str
    namespace: str


class ApproverRef(BaseModel):
    role: Optional[str] = None
    user: Optional[str] = None
    group: Optional[str] = None


class ClientContext(BaseModel):
    ipAddress: Optional[str] = None
    userAgent: Optional[str] = None


class ApprovalDecision(BaseModel):
    action: str
    decidedBy: str
    decidedAt: str
    reason: Optional[str] = None
    clientContext: Optional[ClientContext] = None


class ToolApprovalRequestStatus(BaseModel):
    phase: Optional[str] = None
    observedGeneration: Optional[int] = None
    requestedAt: Optional[str] = None
    decision: Optional[ApprovalDecision] = None
    approvalDuration: Optional[str] = None
    conditions: Optional[List[Any]] = None


class ToolApprovalResponse(BaseModel):
    name: str
    namespace: str
    queryRef: QueryReference
    toolCalls: List[ToolCallInfo]
    phase: Optional[str] = None
    creationTimestamp: Optional[str] = None


class ToolApprovalListResponse(BaseModel):
    items: List[ToolApprovalResponse]
    count: int


class ToolApprovalDetailResponse(BaseModel):
    name: str
    namespace: str
    queryRef: QueryReference
    toolCalls: List[ToolCallInfo]
    timeout: Optional[str] = None
    onTimeout: Optional[str] = None
    approvers: Optional[List[ApproverRef]] = None
    reasonRequired: Optional[bool] = None
    executionContext: ExecutionContext
    status: Optional[ToolApprovalRequestStatus] = None
    metadata: Optional[dict] = None


class ApprovalDecisionRequest(BaseModel):
    action: str = Field(..., pattern="^(approved|rejected)$")
    reason: Optional[str] = None


class ApprovalDecisionResponse(BaseModel):
    name: str
    namespace: str
    phase: str
    decision: ApprovalDecision

File: services/ark-dashboard/ark-dashboard/lib/services/tool-approvals.ts
import { apiClient } from '@/lib/api/client';

export type ToolCallInfo = {
  id: string;
  name: string;
  type: string;
  arguments: string;
  description?: string;
  agentReasoning?: string;
};

export type QueryReference = {
  name: string;
  namespace: string;
};

export type ApprovalDecision = {
  action: 'approved' | 'rejected';
  respondedBy: string;
  respondedAt: string;
  reason?: string;
};

export type ToolApprovalStatus = {
  phase: 'pending' | 'approved' | 'rejected' | 'expired';
  decision?: ApprovalDecision;
  requestedAt?: string;
};

export type ToolApproval = {
  name: string;
  namespace: string;
  queryRef: QueryReference;
  toolCalls: ToolCallInfo[];
  phase?: string;
  creationTimestamp?: string;
};

export type ToolApprovalDetail = ToolApproval & {
  timeout?: string;
  onTimeout?: string;
  reasonRequired?: boolean;
  status?: ToolApprovalStatus;
};

export type ToolApprovalListResponse = {
  items: ToolApproval[];
  count: number;
};

export type ApprovalDecisionRequest = {
  action: 'approved' | 'rejected';
  reason?: string;
};

export type ApprovalDecisionResponse = {
  name: string;
  namespace: string;
  phase: string;
  decision: ApprovalDecision;
};

export const toolApprovalsService = {
  async list(namespace?: string): Promise<ToolApprovalListResponse> {
    const params = namespace ? `?namespace=${namespace}` : '';
    return apiClient.get<ToolApprovalListResponse>(
      `/api/v1/tool-approvals${params}`,
    );
  },

  async listPending(namespace?: string): Promise<ToolApprovalListResponse> {
    const params = namespace ? `?namespace=${namespace}` : '';
    return apiClient.get<ToolApprovalListResponse>(
      `/api/v1/tool-approvals/pending${params}`,
    );
  },

  async get(name: string, namespace?: string): Promise<ToolApprovalDetail> {
    const params = namespace ? `?namespace=${namespace}` : '';
    return apiClient.get<ToolApprovalDetail>(
      `/api/v1/tool-approvals/${name}${params}`,
    );
  },

  async submitDecision(
    name: string,
    decision: ApprovalDecisionRequest,
    namespace?: string,
  ): Promise<ApprovalDecisionResponse> {
    const params = namespace ? `?namespace=${namespace}` : '';
    return apiClient.post<ApprovalDecisionResponse>(
      `/api/v1/tool-approvals/${name}/decision${params}`,
      decision,
    );
  },

  async approve(
    name: string,
    reason?: string,
    namespace?: string,
  ): Promise<ApprovalDecisionResponse> {
    return this.submitDecision(name, { action: 'approved', reason }, namespace);
  },

  async reject(
    name: string,
    reason?: string,
    namespace?: string,
  ): Promise<ApprovalDecisionResponse> {
    return this.submitDecision(name, { action: 'rejected', reason }, namespace);
  },

  async delete(name: string, namespace?: string): Promise<void> {
    const params = namespace ? `?namespace=${namespace}` : '';
    await apiClient.delete(`/api/v1/tool-approvals/${name}${params}`);
  },
};
